Fix invisible sprite bodies

reaperrr opened the pull request:
This fixes 
- the problem of sprite bodies being invisible when they require some condition (RA V2 launcher, several TS units)
- several other yaml-side regressions (some actors were missing the `AutoRenderSize` trait, this refactor now makes that trait obsolete again)

This PR removes `AutoRenderSize` again (and renames `CustomRenderSize` to `CustomRenderBounds`), but first and formost it now refactors the way actor render bounds are calculated. Instead of caching them once in the Actor constructor, they are now calculated when they're called.
Additionally, now the RenderBounds merge all (active) bounds on calculation, instead of simply taking the first, which fixes

Fixes #14449.
Fixes #14462.
Fixes #6987 (for bibs, in any case, `WithShadow` will need its own fix).

penev92 commented:
Left my two cents on the issue. That should cover both "what exactly is causing it" as well as some suggestions on how to fix it.

reaperrr committed:
Revert introduction of AutoRenderSize

Fix DetermineSelectableBounds

Ignoring IAutoSelectionSize is a bug, and returning RendeBounds when an actor lacks SelectableInfo(.Bounds) was at least questionable.
Since we have separate traits for them now, we should keep selecting and rendering more strictly apart since such a fallback might be unintuitive.

Reason for AutoRenderSize reversal: The only reason why it was introduced to begin with was that multiple bounds calculating traits were not supported and TS tunnel tops needed custom bounds.
However, not adding the interface to the sprite bodies directly had too many negative side-effects in terms of regressions and required checks and yaml changes.
Instead, the IAutoRenderSize interface now has a boolean CustomOverride boolean that if true, gives that trait priority over non-override traits.
This allows to add the interface to sprite bodies directly, making the yaml side more simple and robust, while allowing to override their auto-calculation via CustomRenderSize.

reaperrr commented:
Updated.

reaperrr committed:
Rename CustomRenderSize -> CustomRenderBounds

We haven't shipped a pt/release with it yet, so better to do this rename now.

reaperrr committed:
Remove IRenderBounds and use IRender instead

Turned Actor.RenderBounds into property as well, to calculate render bounds at call time.

Actor.RenderBounds are now pretty much defined via RenderSprites and RenderVoxels, although supporting more IRender-ables is now possible and fairly easy (the renderable only needs to return a non-empty Rectangle via AutoRenderBounds).

reaperrr commented:
Updated again. The last 3 commits are only renaming commits, though.

Reviewing this PR commit by commit is strongly recommended.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/Render/RenderSprites.cs:L254: [SP2000] Invalid spacing at the end of the line.

reaperrr commented:
Fixed.

reaperrr committed:
Make RenderSprites return aggregated anim bounds

reaperrr committed:
Use FrozenActor.Actor bounds directly

FA.RenderBounds and FA.SelectableBounds were just using the Actor's values anyway, so lets remove them in favor of using the FA.Actor's bounds directly.

reaperrr committed:
Perform !Rectangle.IsEmpty check in ScreenMap instead of externally

This makes things both simpler and more robust.

reaperrr commented:
Updated.

This now pretty much implements everything @pchote suggested on IRC, except I didn't find an example anf forgot to adress the pixel position fix required for the calculations in RenderSprites/RenderVoxels.

I'll try to adress that tomorrow.
@pchote If you remember or find where you saw that example you mentioned in the meantime, let me know.

pchote commented:
The `CustomRenderBounds` trait will be completely redundant once the `IRender` `Bounds` aggregation is working correctly.  Can we please reorder the PR to do the `IRender` and aggregation stuff first, and then adjust the reverts and renames to simply delete all the Render stuff?

pchote commented on OpenRA.Mods.Common/Traits/Carryall.cs:
```diff
@@ -192,6 +192,12 @@ IEnumerable<IRenderable> IRender.Render(Actor self, WorldRenderer wr)
 
+		Rectangle IRender.AutoRenderBounds(Actor self)
+		{
+			// Actor.RenderBounds unions all non-empty sprite bounds, so this wouldn't have an effect on the final bounds anyway as RenderSprites/RenderBounds take care of that
```
Correct me if i'm wrong, but `RenderSprites` doesn't know anything about the carried actor's renderables, so it won't account for those when returning the carryall's own render bounds.  This will need to aggregate over the previewRenderable's bounds in the same way that `RenderSprites` will.

pchote commented on OpenRA.Mods.Common/Traits/Contrail.cs:
```diff
@@ -71,6 +71,12 @@ IEnumerable<IRenderable> IRender.Render(Actor self, WorldRenderer wr)
+		Rectangle IRender.AutoRenderBounds(Actor self)
+		{
+			// Actor.RenderBounds unions all non-empty sprite bounds, so this wouldn't have an effect on the final bounds anyway as RenderSprites/RenderBounds take care of that
+			return Rectangle.Empty;
```
Likewise here, if we want the contrails to count towards the ScreenMap bounds (which we do... in the long term at least... if we don't want them to be cut off when the actor's body is outside the screen).

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderSprites.cs:
```diff
@@ -236,19 +237,38 @@ public static string NormalizeSequence(Animation anim, DamageState state, string
-				&& b.Animation.Animation.CurrentSequence != null)
-					.Select(a => (a.Animation.Animation.Image.Size.XY * info.Scale).ToInt2())
-					.FirstOrDefault();
+			var visibleAnims = anims.Where(b => b.IsVisible && b.Animation.Animation.CurrentSequence != null).ToArray();
```
This is tricky, because changing the visible animations or the sequence of currently visible animations can change the screen bounds of the actor.  We will need to adopt @penev92's idea for notifying/triggering a bounds recalculation when the visible animations change.

pchote commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -96,14 +96,15 @@ public Damage(int damage)
+	public interface IRender
+	{
+		IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr);
+		Rectangle AutoRenderBounds(Actor self);
```
I think this would be best defined as `IEnumerable<Rectangle> ScreenBounds(Actor self, WorldRenderer wr)`, and then have the implementing traits return the screen-space rectangles of each animation.  This then avoids N different places having to aggregate over the rectangles that they "own".

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderSprites.cs:
```diff
@@ -236,19 +237,38 @@ public static string NormalizeSequence(Animation anim, DamageState state, string
 
-		// Required by WithSpriteBody and WithInfantryBody
-		public int2 AutoRenderSize(Actor self)
+		Rectangle IRender.AutoRenderBounds(Actor self)
```
This could then become something like
```csharp
IEnumerable<Rectangle> ScreenBounds(Actor self, WorldRenderer wr)
{
	foreach (var a in anims)
	{
		if (!a.IsVisible)
			continue;

		foreach (var r in a.Animation.Render(self, wr, null, info.Scale))
		{
			if (!(r is SpriteRenderable))
				continue;

			// NOTE: This could be simplified if we define ScreenBounds on IRenderable
			yield return ((SpriteRenderable)r).ScreenBounds(wr);
		}
	}
}
```

pchote commented:
My [renderbounds](https://github.com/pchote/OpenRA/commits/renderbounds) branch now nicely solves all our issues with invisible actors, bibs, shadows etc.  Unfortunately fixing the actor bounds necessarily breaks the mouseover areas (which wrongly use the actor bounds).

I'm going to need to address my annoying prescient request from #13899 about untangling the selection + selection overlay bounds from the render bounds before I can PR this.

pchote commented:
Superseded by #14482.
